### PR TITLE
Feature/angular modules services routing

### PR DESCRIPTION
**Pull request by AnatoliRub**
**Task scope:**

**404 page**

- If the user entered an incorrect URL which is missing in the router configuration, the app should redirect him to the 404 page

**Login page**

- When the user isn't logged in, the app should prevent him/her from navigating to any pages other than the login page. A router guard would be helpful to implement such behavior.

- There’s no user validation at this stage. You can enter arbitrary credentials and invoke login using an appropriate service.
- Create a login service to enable the required login functionality.

- Once the user has submitted the form, an appropriate method should be called in the login service that saves a fake auth token in localStorage. Then, the user should be redirected to the Main page. If the user gets logged out, he/she should be redirected to the initial login page and the auth token should be deleted from localStorage.

- The login button in the Header component could remain inactive for now, because implementing its functionality requires some RxJS-based features (we’ll deal with it in the upcoming lesson and corresponding task). In order to implement the logout functionality, you can create a separate button that calls the logout method of the login service.

**Main page**

- Move the state and its methods from the Board component to a dedicated service.

- By clicking the More button, the Detailed information page should appear

**Detailed information page**

- Should contain information about the selected video.

- To determine which video is selected, pass its id as a route param

### CODE REVIEW

File: youtube-client/src/app/auth/guards/auth.guard.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import { RoutesPath } from 'src/app/routes.enum';
+import { AuthService } from '../services/auth.service';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class AuthGuard implements CanActivate {
+  constructor(private authService: AuthService, private router: Router) {}
+
+  canActivate(): Promise<boolean> {
```
Comment (StanevichIvan) [inline youtube-client/src/app/auth/guards/auth.guard.ts:12]: might we could rewrite it in async/await style?

File: youtube-client/src/app/auth/pages/login-page/login-page.component.html [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    </div>
+
+    <div class="form__item">
+      <label class="form__item-label" for="password">password:</label>
+      <input class="form__item-input" type="password" id="password" formControlName="password" />
+      <div class="validation" *ngIf="password?.invalid && password?.touched">
+        <small class="validation__item" *ngIf="password?.errors?.['required']"
+          >This field is required.</small
+        >
+        <small class="validation__item" *ngIf="password?.errors?.['minlength']"
```
Comment (StanevichIvan) [inline youtube-client/src/app/auth/pages/login-page/login-page.component.html:27]: Please think about how you can create reusable error component

File: youtube-client/src/app/core/components/filtering-criteria/filtering-criteria.component.ts [hunk 3]
```diff
@@ -20,7 +21,7 @@ export class FilteringCriteriaComponent implements OnDestroy {
 
   @Input() filterWord = '';
 
-  @Output() filterData = new EventEmitter<IFilterData>();
+  constructor(private searchService: SearchServiceService) {}
```
Comment (StanevichIvan) [inline youtube-client/src/app/core/components/filtering-criteria/filtering-criteria.component.ts:24]: might private readonly

File: youtube-client/src/app/youtube/pages/detailed-page/detailed-page.component.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    private cardService: CardService,
+    private router: Router,
+  ) {}
+
+  goBack() {
+    this.router.navigate([RoutesPath.Youtube]);
+  }
+
+  ngOnInit(): void {
+    this.route.params.subscribe((params: Params) => {
```
Comment (StanevichIvan) [inline youtube-client/src/app/youtube/pages/detailed-page/detailed-page.component.ts:29]: Where do we unsubscribe?

File: youtube-client/src/app/core/components/filtering-criteria/filtering-criteria.component.ts (hunk 3 above)
Comment (AnatoliRub) [reply youtube-client/src/app/core/components/filtering-criteria/filtering-criteria.component.ts]: a0aebae

File: youtube-client/src/app/auth/guards/auth.guard.ts (hunk 1 above)
Comment (AnatoliRub) [reply youtube-client/src/app/auth/guards/auth.guard.ts]: 2e3f2ca

File: youtube-client/src/app/youtube/pages/detailed-page/detailed-page.component.ts (hunk 4 above)
Comment (AnatoliRub) [reply youtube-client/src/app/youtube/pages/detailed-page/detailed-page.component.ts]: ee4e001

File: youtube-client/src/app/auth/components/error/error.component.scss [hunk 5]
```diff
@@ -0,0 +1,8 @@
+.validation {
+  padding-left: rem(6);
+
+  &__item {
+    color: var(--red);
+    margin-right: rem(3);
+  }
+}
```
Comment (StanevichIvan) [inline youtube-client/src/app/auth/components/error/error.component.scss:8]: Is it prettier issue?

Empty line id required in the file end

File: youtube-client/src/app/auth/pages/login-page/login-page.component.html (hunk 2 above)
Comment (AnatoliRub) [reply youtube-client/src/app/auth/pages/login-page/login-page.component.html]: bfb8379
22d68f3
c80deb0
5287b20

File: youtube-client/src/app/auth/components/error/error.component.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  selector: 'app-error',
+  templateUrl: './error.component.html',
+  styleUrls: ['./error.component.scss'],
+})
+export class ErrorComponent {
+  @Input() control?: AbstractControl | null;
+
+  @Input() showError?: boolean;
+
+  chooseMessage() {
```
Comment (StanevichIvan) [inline youtube-client/src/app/auth/components/error/error.component.ts:15]: please add return type.
and might `getMessage` or something similar is an option for the component

File: youtube-client/src/app/auth/validators/onlyLetters.validator.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import { AbstractControl, ValidationErrors } from '@angular/forms';
+
+export function OnlyLetters(control: AbstractControl): ValidationErrors | null {
```
Comment (StanevichIvan) [inline youtube-client/src/app/auth/validators/onlyLetters.validator.ts:3]: why do we have not a constructor function with first capital letter

**Review by StanevichIvan — APPROVED** [review]

File: youtube-client/src/app/auth/components/error/error.component.scss (hunk 5 above)
Comment (AnatoliRub) [reply youtube-client/src/app/auth/components/error/error.component.scss]: I'm sorry. I had a several es-lint mistakes about line's length. Its were more than 140 characters and prettier didn't work.
d9ee807

File: youtube-client/src/app/auth/components/error/error.component.ts (hunk 6 above)
Comment (AnatoliRub) [reply youtube-client/src/app/auth/components/error/error.component.ts]: 84bc70b

File: youtube-client/src/app/auth/validators/onlyLetters.validator.ts (hunk 7 above)
Comment (AnatoliRub) [reply youtube-client/src/app/auth/validators/onlyLetters.validator.ts]: 40f111d

**Pull request merged**